inference-engine/src/transformations/src/transformations/common_optimizations/convert_to_unsigned_nms_gather.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/common_optimizations/convert_to_unsigned_nms_gather.hpp"
+#include <ngraph/opsets/opset8.hpp>
+#include <ngraph/rt_info.hpp>
+#include <queue>
+#include <memory>
+#include "itt.hpp"
+
+using namespace ngraph;
+using namespace std;
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::ConvertToUnsignedNmsGather, "ConvertToUnsignedNmsGather", 0);
+
+
+bool ngraph::pass::ConvertToUnsignedNmsGather::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    MATCHER_SCOPE(MarkNMSGather);
+    bool res = false;
+    for (const auto& op : f->get_ordered_ops()) {
+        // without dynamic_cast since not all NMS versions are inherited from the same base nms
+        if (strcmp(op->get_type_name(), opset8::NonMaxSuppression::type_info.name) != 0)
+            continue;
+        convert_gather_indices_to_unsigned(get_NmsGather_destinations(op));
+        res = true;
+    }
+    return res;
+}
+
+NodeVector get_NmsGather_destinations(const shared_ptr<Node>& nms_node) {
+    NodeVector res;
+    set<string> skip_node_types = {
+            opset8::Squeeze::type_info.name,
+            opset8::Unsqueeze::type_info.name,
+            opset8::Reshape::type_info.name,
+            opset8::Broadcast::type_info.name,
+            opset8::StridedSlice::type_info.name,
+            opset8::VariadicSplit::type_info.name,
+            opset8::Concat::type_info.name,
+            opset8::Convert::type_info.name
+    };
+
+    // BFS search
+    std::unordered_set<shared_ptr<Node>> visited;
+    std::queue<shared_ptr<Node>> stack;
+    stack.push(nms_node->output(0).get_node_shared_ptr());
+
+    while (!stack.empty()) {
+        shared_ptr<Node> curr = stack.front();
+        visited.insert(curr);
+        stack.pop();
+
+        for (const auto& next : curr->get_users()) {
+            if (!visited.count(next) && skip_node_types.count(next->get_type_name())) {
+                stack.push(next);
+            } else if (!visited.count(next) && strcmp(next->get_type_name(), opset8::Gather::type_info.name) == 0) {
+                // if next goes into Gather indices input
+                if (next->input_value(1).get_node_shared_ptr() == curr) {
+                    res.emplace_back(next);
+                }
+            }
+        }
+    }
+    return res;
+}
+
+void convert_gather_indices_to_unsigned(const NodeVector& gather_nodes) {
+    for (const auto& gather : gather_nodes) {
+        auto indices = gather->input_value(1);
+
+        auto out_type = element::Type_t::u32;
+        if (indices.get_element_type() == element::Type_t::i64)
+            out_type = element::Type_t::u64;
```

Comment:
Are we sure that `Convert` behaves as we expected for `-1` convertion into `uint_max`? I found in Convert specification: `The result of unsupported conversions is undefined, e.g. conversion of negative signed integer value to any unsigned integer type.`

Reply:
You raised a good question! In practice, when you convert to unsigned locally in C++ -1 -> uint32_max: the same bits of -1 if we interpret as unsigned the it's uint32_max.

The whole picture indeed is quite involved. During inference unsigned types are converted to signed 
CPU https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp#L129

VPU https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/vpu/graph_transformer/src/frontend/frontend.cpp#L195

so indeed in plugins now we will have -1 again, but this soon will be changed in a such way that plugins will check 
if indices data type is unsigned then -> run Gather in the old v7 way, we discussed this locally with plugins. @slyalin can confirm.

But thank you for pointing to the specification, we should have aligned specification and implementation.

Reply:
Updated `Convert` specification to match c++ integer types conversion standard (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4659.pdf, page 89)
![image](https://user-images.githubusercontent.com/5703039/127831857-61fae3e5-9917-4d0e-bc15-2884e12bcfd9.png)


Since no new behavior have been introduced but only concretized behavior for conversion into unsigned no new version of operation is needed.